tests/bmtestspolar/symfixedrankpolarfactory.m

```diff
@@ -0,0 +1,215 @@
+function M = symfixedrankpolarfactory(m, k)
+% Manifold of m-by-m symmetric positive definite matrices of rank k with polar quotient geometry.
+%
+% function M = symfixedrankpolarfactory(m, k)
+%
+% The first-order geometry follows the balanced quotient geometry described
+% in the paper,
+% "Linear regression under fixed-rank constraints: a Riemannian approach",
+% G. Meyer, B. Bonnabel and R. Sepulchre, ICML 2011.
+%
+% Paper link: http://www.icml-2011.org/papers/350_icmlpaper.pdf.
+%
+% The second-order geometry follows from the paper
+% "Fixed-rank matrix factorizations and Riemannian low-rank optimization",
+% B. Mishra, R. Meyer, B. Bonnabel and R. Sepulchre,
+% Computational Statistics, 29(3 - 4), pp. 591 - 621, 2014.
+%
+% A point X on the manifold is represented as a structure with two
+% fields: U and B. The matrix U (mxk) is orthonormal,
+% while the matrix B (kxk) is a symmetric positive definite full rank
+% matrix.
+%
+% Tangent vectors are represented as a structure with two fields: U and B.
+%
+%
+% For first-order geometry, please cite the Manopt paper as well as the research paper:
+%     @InProceedings{meyer2011linear,
+%       Title        = {Linear regression under fixed-rank constraints: a {R}iemannian approach},
+%       Author       = {Meyer, G. and Bonnabel, B. and Sepulchre, R.},
+%       Booktitle    = {{28th International Conference on Machine Learning}},
+%       Year         = {2011},
+%       Organization = {{ICML}}
+%     }
+% For second-order geometry, please cite the Manopt paper as well as the research paper:
+%     @Article{mishra2014fixedrank,
+%       Title   = {Fixed-rank matrix factorizations and {Riemannian} low-rank optimization},
+%       Author  = {Mishra, B. and Meyer, G. and Bonnabel, B. and Sepulchre, R.},
+%       Journal = {Computational Statistics},
+%       Year    = {2014},
+%       Number  = {3-4},
+%       Pages   = {591--621},
+%       Volume  = {29},
+%       Doi     = {10.1007/s00180-013-0464-z}
+%     }
+%
+%
+% See also symfixedrankYYfactory sympositivedefinitefactory
+    
+    % This file is part of Manopt: www.manopt.org.
+    % Original author: Bamdev Mishra, April 05, 2019.
+    % Contributors:
+    % Change log:
+    %
+    
+    M.name = @() sprintf('UBU'' quotient manifold of %dx%d matrices of rank %d', m, m, k);
+    
+    M.dim = @() m*k - k*(k-1)/2;
+    
+    % Choice of the metric on the orthnormal space is motivated by the symmetry present in the
+    % space. The metric on the positive definite space is its natural metric.
+    M.inner = @(X, eta, zeta) eta.U(:).'*zeta.U(:)  ...
+        + trace( (X.B\eta.B) * (X.B\zeta.B) );
+    
+    M.norm = @(X, eta) sqrt(M.inner(X, eta, eta));
+    
+    M.dist = @(x, y) error('symfixedrankpolarfactory.dist not implemented yet.');
+    
+    M.typicaldist = @() 10*k;
+    
+    skew = @(X) .5*(X-X');
+    symm = @(X) .5*(X+X');
+    stiefel_proj = @(U, H) H - U*symm(U'*H);
+    
```

Comment:
Commonly useful tools would be something to go back and forth between a pair (B, U) and the corresponding matrix UBU' --- it's rarely used as is for actual fast code, but it's very convenient for prototyping. I suggest adding functions M.matrix2pair and M.pair2matrix.

Here is an example of what that looks like for fixedrankembeddedfactory:


    % It is sometimes useful to switch between representation of matrices
    % as triplets or as full matrices of size m x n. The function to
    % convert a matrix to a triplet, matrix2triplet, allows to specify the
    % rank of the representation. By default, it is equal to k. Omit the
    % second input (or set to inf) to get a full SVD triplet (in economy
    % format). If so, the resulting triplet does not represent a point on
    % the manifold.
    M.matrix2triplet = @matrix2triplet;
    function X_triplet = matrix2triplet(X_matrix, r)
        if ~exist('r', 'var') || isempty(r) || r <= 0
            r = k;
        end
        if r < min(m, n)
            [U, S, V] = svds(X_matrix, r);
        else
            [U, S, V] = svd(X_matrix, 'econ');
        end
        X_triplet.U = U;
        X_triplet.S = S;
        X_triplet.V = V;
    end
    M.triplet2matrix = @triplet2matrix;
    function X_matrix = triplet2matrix(X_triplet)
        U = X_triplet.U;
        S = X_triplet.S;
        V = X_triplet.V;
        X_matrix = U*S*V';
    end

Reply:
Done